tree/tree/src/TBasket.cxx

```diff
@@ -724,13 +724,56 @@ void TBasket::AdoptBuffer(TBuffer *user_buffer)
    fBufferRef = user_buffer;
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Reset the read basket TBuffer memory allocation if needed.
+///
+/// This allows to reduce the number of memory allocation while avoiding to 
+/// always use the maximum size.
+
+void TBasket::ReadResetBuffer(Int_t basketnumber)
+{
+    // Downsize the buffer if needed.
+
+   const auto maxbaskets = fBranch->GetMaxBaskets();
+   if (!fBufferRef || basketnumber >= maxbaskets)
+      return;
+
+   Int_t curSize = fBufferRef->BufferSize();
+
+   const Float_t target_mem_ratio = fBranch->GetTree()->GetTargetMemoryRatio();
+   const auto basketbytes = fBranch->GetBasketBytes();
+
+   Int_t max_size = basketbytes[basketnumber];
+   for(Int_t b = basketnumber + 1; (b < maxbaskets) && (b < (basketnumber+10)); ++b) {
+      max_size = std::max(max_size, basketbytes[b]);
+   }
+
+   Float_t cx = 1;
+   if (fBranch->GetZipBytes())
+      cx = (fBranch->GetTotBytes()+0.00001)/fBranch->GetZipBytes();
+
+   Int_t target_size = static_cast<Int_t>(cx * target_mem_ratio * Float_t(max_size));
+
+   if (target_size && (curSize > target_size)) {
+      /// Only reduce the size if significant enough?
+      Int_t newSize = max_size + 512 - max_size % 512; // Wiggle room and alignment, as above.
+      // We only bother with a resize if it saves 8KB (two normal memory pages).
+      if ((newSize <= curSize - 8 * 1024) &&
+         (static_cast<Float_t>(curSize) / static_cast<Float_t>(newSize) > target_mem_ratio))
+      {
+         //Info("Reset", "Resizing %d to %d bytes (was %d); next three sizes were [%d, %d, %d]. cx=%f ratio=%f max_size=%d", basketnumber, newSize, curSize,
+         //     basketbytes[basketnumber], (basketnumber+1) < maxbaskets ? basketbytes[basketnumber+1] : 0, (basketnumber+2) < maxbaskets ? basketbytes[basketnumber+2] : 0, cx, target_mem_ratio, max_size);
+         fBufferRef->Expand(newSize, kFALSE);
```

Comment:
We are not including this in `R__TRACK_BASKET_ALLOC_TIME`?

Reply:
added.